### PR TITLE
Separate prognostic and boundary tensors end-to-end

### PR DESCRIPTION
**Pull request by alxmrs**
Thread prognostic and boundary inputs as separate tensors through the data pipeline, stepper, base model, and all model implementations. Previously they were concatenated along the channel dim before reaching the model; now each model's forward_once receives (prog, boundary, ctx).

FOMO concatenates before its single-stream encoder (the dual-perceiver encoder that enables cross-resolution fusion lands in a follow-up). Samudra and FOMini concatenate inside their forward_once as before.

This is #681 broken up into a stack of three small PRs. 1/3

### CODE REVIEW

**Comment by alxmrs** [comment]
@codex may I have a review of this PR?

File: src/ocean_emulators/models/base.py [hunk 1]
```diff
@@ -101,7 +93,8 @@ def inference(
         num_steps=None,
         epoch=None,
     ) -> ModelInferenceOutput:
-        out_shape = (num_steps, *dataset[0][1].shape[1:])
+        # `dataset[idx]` returns `(prog, boundary, label)`.
```
Comment (jder) [inline src/ocean_emulators/models/base.py:96]: <3

File: src/ocean_emulators/datasets.py [hunk 2]
```diff
@@ -154,20 +155,28 @@ def get_target_time(self, start_step: int, num_steps: int):
             )
         )
 
-    def merge_prognostic_and_boundary(self, prognostic: torch.Tensor, step: int):
+    def get_boundary_for_prognostic(
+        self, prognostic: Prognostic, step: int
+    ) -> tuple[Prognostic, Boundary]:
```
Comment (jder) [inline src/ocean_emulators/datasets.py]: I think this is kind of a weird method now? Should it be `get_boundary(step: int)`? I don't really think we need the `to` in here but if we do I'd rather have callers do `get_boundary(step).to(prognostic.device)`, I think?

File: src/ocean_emulators/models/base.py [hunk 3]
```diff
@@ -113,26 +106,18 @@ def inference(
...
-                pred = (
-                    input_tensor[
-                        0,
-                        : self.out_channels,
-                    ].to(  # Residuals on last state in input
-                        device=get_device()
-                    )
-                    + decodings
-                )
+                pred = prog_tensor.to(device=get_device()) + decodings
```
Comment (jder) [inline src/ocean_emulators/models/base.py]: We seem to have lost the `[0, ...` part from the old code. IIUC the prog_tensor has a batch dimension which we're trying to get rid of here before adding decodings. I guess we don't have inference tests that would catch this? Or am I missing something?

File: src/ocean_emulators/datasets.py [hunk 4]
```diff
@@ -357,44 +366,40 @@ def pin_memory(self):
...
-    autoregressive rollout in the training and inference process.
-
-    Constraint: The `Input` tensor is a combination of (flattened) prognostic variables (at all depth levels) and
-    boundary forcings. The top `num_prognostic_channels` number of channels must be prognostic variables whereas the
-    remaining bottom channels are boundary forcings.
+    A single batch contains multiple steps worth of ``Example`` entries, each
+    of which is a ``(prognostic_input, boundary_input, label)`` triple. The
+    prognostic and boundary tensors are carried *separately* (they are no
+    longer concatenated along the channel dim) so that the encoder can fuse
+    them at the token level, potentially at different spatial resolutions.
```
Comment (jder) [inline src/ocean_emulators/datasets.py]:
```suggestion
    prognostic and boundary tensors are carried separately because the FOMO model
    encodes them separately (Samudra just concatenates them later).
```

File: src/ocean_emulators/datasets.py [hunk 5]
```diff
@@ -622,43 +636,27 @@ def to_train_data(
...
-            boundary_steps = boundary.normalize_and_mask(
-                self.normalize_before_mask, self.masked_fill_value
-            )
-            boundary_steps = flatten_dims(boundary_steps)
-            return torch.cat((prognostic_steps, boundary_steps), dim=1)
-
-        return prognostic_steps
+        return rearrange(
+            steps, "batch time variable lat lon -> batch (time variable) lat lon"
+        )
```
Comment (jder) [inline src/ocean_emulators/datasets.py:649]: Nice cleanup!

File: src/ocean_emulators/stepper.py [hunk 6]
```diff
@@ -20,112 +20,112 @@
...
-                coords=coords,
-                hist=inf_aggregator.hist,
-                model_path=model_path,
-                time_chunk_size=chunk_size,
+    outs = model(batch)[0]
+    loss_per_channel = loss_fn(outs, label, batch.ctx)
+    loss = torch.mean(loss_per_channel)
+    # `input_data` in ValBatchOutput is used for val visualization; pass the
+    # channel-concatenated tensor for continuity with existing consumers.
+    # We only ever mix scales across boundary and prognostic during inference.
```
Comment (jder) [inline src/ocean_emulators/stepper.py]: Does "inference" mean "forward passes" here?

File: src/ocean_emulators/stepper.py [hunk 7]
```diff
@@ -20,112 +20,112 @@
...
+    input_data = torch.cat((prognostic, boundary), dim=1)
+    return ValBatchOutput(loss, loss_per_channel, input_data, label, outs, batch.ctx)
+
+
+@torch.no_grad()
+def inference(
```
Comment (jder) [inline src/ocean_emulators/stepper.py]: Maybe worth a slightly more descriptive name now? eg `run_rollout`?

**Review by jder — APPROVED** [review]
A few suggestions but generally LGTM! Love the getting rid of Stepper. I kinda wish Claude did not leave so many vague comments about matching "existing behavior" but 🤷

File: src/ocean_emulators/datasets.py (hunk 5 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py:649]: Thanks, I was happy about this too.

File: src/ocean_emulators/stepper.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/stepper.py]: hmm, yes, you are right, I'm confusing which AR process is done here.

Comment (alxmrs) [reply src/ocean_emulators/stepper.py]: Wait, looking closer, I do think that it is during eval/inference. Maybe I have the wrong mental model here. To be safe, I'll invert the negative of the sentence and say what I know to be true.

File: src/ocean_emulators/models/base.py (hunk 3 above)
Comment (alxmrs) [reply src/ocean_emulators/models/base.py]: I know this is covered by tests because when I removed the `unexpand(0)` a test failed. I tried two variants (the status quo and a adhering to the original version) and both versions worked. Merging this version.

Comment (jder) [reply src/ocean_emulators/models/base.py]: @fomo-bot can you help me understand why this version works given my above comments?

Comment (jder) [reply src/ocean_emulators/models/base.py]: @fomo-bot ok, can you please try again to help me understand why this version works given my above comments?
